linchpin/hooks/__init__.py

```diff
@@ -0,0 +1,232 @@
+          type: ansible
+          actions:
+            - playbook: test_playbook.yaml
+              vars: test_var.yaml
+              extra_vars: { 'testvar': 'world'}
+
+'''
+
+import pprint
+import ast
+import sys
+
+from linchpin.hooks.action_managers import ACTION_MANAGERS
+from linchpin.exceptions import ActionManagerError
+
+class ActionBlockRouter(object):
+    '''
+    proxy pattern implementation for fetching actionmanagers by name
+
+    '''
+    def __init__(self, name, *args, **kwargs):
+
+        self.__implementation = self.__get_implementation(name)(name, *args, **kwargs)
+
+    def __getattr__(self, name):
+
+        return getattr(self.__implementation, name)
+
+    def __get_implementation(self, class_name):
+        '''
+        Fetches implementation of lcass from linchpin.hooks.action_managers
+        :param name: action manager class name  
+        '''
+
+        action_class = ACTION_MANAGERS.get(class_name, None)
+        if action_class == None:
+            raise ActionManagerError('Action Class {0} not found '.format(class_name))
+        return action_class
+
+
+class LinchpinHooks(object):
+
+    def __init__(self, api):
+        '''
+        FIXME
+        '''
+
+        self.api = api
+        self.api.bind_to_hook_state(self.run_hooks)
+
+
+    def prepare_ctx_params(self):
+        '''
+        prepares few context parameters based on the current target_data
+        that is being set. these parameters are based topology name.
+        '''
+
+        topology = self.api.target_data['topology']
+        name = topology.split('/')[-1].split('.')[-2]
+        inv_file = '{0}{1}'.format(name,
+                            self.api.get_cfg('extensions', 'inventory'))
+        res_file = '{0}{1}'.format(name,
+                            self.api.get_cfg('extensions', 'resource'))
+        inv_file = '{0}/{1}'.format(
+           self.api.target_data['extra_vars']['default_inventories_path'],
+           inv_file
+           )
+        res_file = '{0}/{1}'.format(
+               self.api.target_data['extra_vars']['default_resources_path'],
+               res_file
+               )
+        self.api.target_data['extra_vars']['inventory_file'] = inv_file
+        self.api.target_data['extra_vars']['resource_file'] = res_file
+
+
+    def run_hooks(self, state, is_global=False):
+
+        '''
+        Run any hooks for the current state
+
+        :param state: hook state (currently, preup, postup,
+        predestroy, postdestroy)
+        :param is_global: whether the hook is global (can be applied to
+        multiple targets)
+        '''
+
+        # FIXME
+        # This statement makes no logical sense.
+        # Why would we prepare context parameters if the state isn't
+        # listed in the method? It doesn't appear as though that
+        # sets any related values in any way, shape, or form.
+        #self.prepare_ctx_params()
+
+        hooks_data = self.api.target_data.get('hooks', None)
```

Comment:
This breaks when there are no hooks for a target. A simple fix would look like this::

  if hooks_data and hooks_data.has_key(str(state)):